File: front-end/src/services/overview/overviewService.js
import { getAllFocuses } from "../focus/focusService";
import { getGitHubAnalytics } from "../github/githubAnalyticsService";
import { getAllGoals } from "../goal/goalService";
import { getAllSessions } from "../session/sessionService";
import { getTasksByFocus } from "../task/taskService";

function toDateKey(value) {
  return new Date(value).toISOString().slice(0, 10);
}

function getLastSevenDays() {
  const today = new Date();

  return Array.from({ length: 7 }, (_, index) => {
    const date = new Date(today);
    date.setDate(today.getDate() - (6 - index));

    return toDateKey(date);
  });
}

function getCurrentStreak(activeDateKeys) {
  let streak = 0;
  const cursor = new Date();

  while (activeDateKeys.has(toDateKey(cursor))) {
    streak += 1;
    cursor.setDate(cursor.getDate() - 1);
  }

  return streak;
}

function buildTaskStats(tasks) {
  const completed = tasks.filter((task) => Boolean(task.status)).length;
  const total = tasks.length;
  const pending = total - completed;

  return {
    completionRate: total > 0 ? Math.round((completed / total) * 100) : 0,
    completed,
    pending,
    total,
  };
}

function buildGoalStats(goals) {
  return {
    completed: goals.filter((goal) => goal.status === "DONE").length,
    discarded: goals.filter((goal) => goal.status === "DISCARDED").length,
    inProgress: goals.filter((goal) => goal.status === "IN_PROGRESS").length,
    pending: goals.filter((goal) => goal.status === "TODO").length,
    total: goals.length,
  };
}

function buildSessionStats(sessions) {
  const totalDuration = sessions.reduce(
    (sum, session) => sum + (session.duration || 0),
    0
  );
  const averageDuration =
    sessions.length > 0 ? Math.round(totalDuration / sessions.length) : 0;
  const lastSevenDays = getLastSevenDays();
  const durationByDate = sessions.reduce((accumulator, session) => {
    if (!session.createdAt) return accumulator;

    const dateKey = toDateKey(session.createdAt);
    accumulator[dateKey] = (accumulator[dateKey] || 0) + (session.duration || 0);

    return accumulator;
  }, {});
  const weeklyActivity = lastSevenDays.map((date) => ({
    date,
    duration: durationByDate[date] || 0,
  }));
  const activeDateKeys = new Set(
    Object.entries(durationByDate)
      .filter(([, duration]) => duration > 0)
      .map(([date]) => date)
  );

  return {
    activeDaysLastSeven: weeklyActivity.filter((day) => day.duration > 0).length,
    averageDuration,
    currentStreak: getCurrentStreak(activeDateKeys),
    total: sessions.length,
    totalDuration,
    weeklyActivity,
  };
}

function buildFocusStats(focuses, tasks) {
  const tasksByFocus = tasks.reduce((accumulator, task) => {
    const focusId = task.idFocus;
    accumulator[focusId] = (accumulator[focusId] || 0) + 1;

    return accumulator;
  }, {});
  const activeFocus = focuses
    .map((focus) => ({
      ...focus,
      taskCount: tasksByFocus[focus.id] || 0,
    }))
    .sort((firstFocus, secondFocus) => secondFocus.taskCount - firstFocus.taskCount)[0];

  return {
    activeFocus: activeFocus?.title || "None",
    total: focuses.length,
  };
}

function buildEmptyGitHubStats() {
  return {
    connected: false,
    username: null,
    publicRepos: 0,
    commitsLastSevenDays: 0,
    commitsLastThirtyDays: 0,
    stacks: [],
    repositories: [],
    frequency: getLastSevenDays().map((date) => ({
      date,
      commits: 0,
    })),
  };
}

export async function getOverviewData() {
  const [focuses, goals, sessions, githubResult] = await Promise.all([
    getAllFocuses(),
    getAllGoals(),
    getAllSessions(),
    getGitHubAnalytics().catch(() => buildEmptyGitHubStats()),
  ]);
  const taskGroups = await Promise.all(
    focuses.map(async (focus) => getTasksByFocus(focus.id))
  );
  const tasks = taskGroups.flat();

  return {
    focuses: buildFocusStats(focuses, tasks),
    goals: buildGoalStats(goals),
    github: githubResult || buildEmptyGitHubStats(),
    sessions: buildSessionStats(sessions),
    tasks: buildTaskStats(tasks),
  };
}

File: back-end/src/main/java/org/danielbreves/backend/service/OverviewService.java
package org.danielbreves.backend.service;

import org.danielbreves.backend.dto.github.GitHubAnalyticsResponseDTO;
import org.danielbreves.backend.dto.overview.OverviewFocusStatsDTO;
import org.danielbreves.backend.dto.overview.OverviewGoalStatsDTO;
import org.danielbreves.backend.dto.overview.OverviewResponseDTO;
import org.danielbreves.backend.dto.overview.OverviewSessionActivityDTO;
import org.danielbreves.backend.dto.overview.OverviewSessionStatsDTO;
import org.danielbreves.backend.dto.overview.OverviewStreakStatsDTO;
import org.danielbreves.backend.dto.overview.OverviewTaskStatsDTO;
import org.danielbreves.backend.entity.Focus;
import org.danielbreves.backend.entity.Goal;
import org.danielbreves.backend.entity.Session;
import org.danielbreves.backend.entity.Task;
import org.danielbreves.backend.entity.User;
import org.danielbreves.backend.entity.enums.GoalStatus;
import org.danielbreves.backend.exception.NotFoundException;
import org.danielbreves.backend.repository.FocusRepository;
import org.danielbreves.backend.repository.GoalRepository;
import org.danielbreves.backend.repository.SessionRepository;
import org.danielbreves.backend.repository.TaskRepository;
import org.danielbreves.backend.repository.UserRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class OverviewService {
    private static final ZoneId ACTIVITY_ZONE = ZoneId.of("America/Sao_Paulo");

    private final FocusRepository focusRepository;
    private final GoalRepository goalRepository;
    private final SessionRepository sessionRepository;
    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final GitHubAnalyticsService gitHubAnalyticsService;

    public OverviewService(
            FocusRepository focusRepository,
            GoalRepository goalRepository,
            SessionRepository sessionRepository,
            TaskRepository taskRepository,
            UserRepository userRepository,
            GitHubAnalyticsService gitHubAnalyticsService
    ) {
        this.focusRepository = focusRepository;
        this.goalRepository = goalRepository;
        this.sessionRepository = sessionRepository;
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.gitHubAnalyticsService = gitHubAnalyticsService;
    }

    public OverviewResponseDTO getOverview(String currentEmail) {
        User user = userRepository.findByEmail(currentEmail)
                .orElseThrow(() -> new NotFoundException("User not found"));
        List<Focus> focuses = focusRepository.findAllByUser(user);
        List<Task> tasks = taskRepository.findAllByFocus_User(user);
        List<Goal> goals = goalRepository.findAllByIdUser(user.getId());
        List<Session> sessions = sessionRepository.findAllByUser(user);
        GitHubAnalyticsResponseDTO github =
                gitHubAnalyticsService.getAnalytics(currentEmail);
        OverviewSessionStatsDTO sessionStats = buildSessionStats(sessions);
        OverviewStreakStatsDTO streakStats = github.connected()
                ? buildGitHubStreakStats(github)
                : new OverviewStreakStatsDTO(
                        sessionStats.activeDaysLastSeven(),
                        sessionStats.currentStreak()
                );

        return new OverviewResponseDTO(
                buildFocusStats(focuses, tasks),
                buildGoalStats(goals),
                github,
                sessionStats,
                streakStats,
                buildTaskStats(tasks)
        );
    }

    private OverviewFocusStatsDTO buildFocusStats(
            List<Focus> focuses,
            List<Task> tasks
    ) {
        Map<Long, Long> tasksByFocus = tasks.stream()
                .collect(Collectors.groupingBy(
                        task -> task.getFocus().getId(),
                        Collectors.counting()
                ));
        String activeFocus = focuses.stream()
                .max((firstFocus, secondFocus) -> Long.compare(
                        tasksByFocus.getOrDefault(firstFocus.getId(), 0L),
                        tasksByFocus.getOrDefault(secondFocus.getId(), 0L)
                ))
                .map(Focus::getTitle)
                .orElse("None");

        return new OverviewFocusStatsDTO(activeFocus, focuses.size());
    }

    private OverviewGoalStatsDTO buildGoalStats(List<Goal> goals) {
        return new OverviewGoalStatsDTO(
                countGoalsByStatus(goals, GoalStatus.DONE),
                countGoalsByStatus(goals, GoalStatus.DISCARDED),
                countGoalsByStatus(goals, GoalStatus.IN_PROGRESS),
                countGoalsByStatus(goals, GoalStatus.TODO),
                goals.size()
        );
    }

    private OverviewSessionStatsDTO buildSessionStats(List<Session> sessions) {
        int totalDuration = sessions.stream()
                .mapToInt(session -> safeDuration(session.getDurationSeconds()))
                .sum();
        int averageDuration = sessions.isEmpty()
                ? 0
                : Math.round((float) totalDuration / sessions.size());
        Map<String, Integer> durationByDate = sessions.stream()
                .filter(session -> session.getCreatedAt() != null)
                .collect(Collectors.toMap(
                        session -> session.getCreatedAt().toLocalDate().toString(),
                        session -> safeDuration(session.getDurationSeconds()),
                        Integer::sum,
                        LinkedHashMap::new
                ));
        List<OverviewSessionActivityDTO> weeklyActivity =
                getLastSevenDays().stream()
                        .map(date -> new OverviewSessionActivityDTO(
                                date,
                                durationByDate.getOrDefault(date, 0)
                        ))
                        .toList();
        Set<String> activeDateKeys = durationByDate.entrySet().stream()
                .filter(entry -> entry.getValue() > 0)
                .map(Map.Entry::getKey)
                .collect(Collectors.toSet());

        return new OverviewSessionStatsDTO(
                (int) weeklyActivity.stream()
                        .filter(day -> day.duration() > 0)
                        .count(),
                averageDuration,
                getCurrentStreak(activeDateKeys),
                sessions.size(),
                totalDuration,
                weeklyActivity
        );
    }

    private OverviewStreakStatsDTO buildGitHubStreakStats(
            GitHubAnalyticsResponseDTO github
    ) {
        Set<String> activeDateKeys = github.frequency().stream()
                .filter(day -> day.commits() > 0)
                .map(day -> day.date())
                .collect(Collectors.toSet());

        return new OverviewStreakStatsDTO(
                (int) github.frequency().stream()
                        .filter(day -> day.commits() > 0)
                        .count(),
                getCurrentStreak(activeDateKeys)
        );
    }

    private OverviewTaskStatsDTO buildTaskStats(List<Task> tasks) {
        int completed = (int) tasks.stream()
                .filter(task -> Boolean.TRUE.equals(task.getStatus()))
                .count();
        int total = tasks.size();
        int pending = total - completed;
        int completionRate = total > 0
                ? Math.round((completed * 100f) / total)
                : 0;

        return new OverviewTaskStatsDTO(
                completed,
                completionRate,
                pending,
                total
        );
    }

    private int countGoalsByStatus(List<Goal> goals, GoalStatus status) {
        return (int) goals.stream()
                .filter(goal -> goal.getStatus() == status)
                .count();
    }

    private int safeDuration(Integer duration) {
        return duration == null ? 0 : duration;
    }

    private List<String> getLastSevenDays() {
        LocalDate today = LocalDate.now(ACTIVITY_ZONE);

        return java.util.stream.IntStream.rangeClosed(0, 6)
                .mapToObj(index -> today.minusDays(6L - index).toString())
                .toList();
    }

    private int getCurrentStreak(Set<String> activeDateKeys) {
        int streak = 0;
        LocalDate cursor = LocalDate.now(ACTIVITY_ZONE);

        if (!activeDateKeys.contains(cursor.toString())) {
            cursor = cursor.minusDays(1);
        }

        while (activeDateKeys.contains(cursor.toString())) {
            streak++;
            cursor = cursor.minusDays(1);
        }

        return streak;
    }
}
